### PR TITLE
Update BoundCall method based on receiver nullability

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

**Comment by cston** [comment]
@dotnet-bot test this please

**Comment by jcouv** [comment]
Fixes https://github.com/dotnet/roslyn/issues/29605?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -51406,5 +51406,112 @@ static void Main()
             var comp = CreateCompilation(source2, new[] { ref1.WithEmbedInteropTypes(true), CSharpRef }, options: WithNonNullTypesTrue(TestOptions.ReleaseExe));
             CompileAndVerify(comp, expectedOutput: "4");
         }
+
+        [WorkItem(29605, "https://github.com/dotnet/roslyn/issues/29605")]
+        [Fact]
+        public void InferMethodReceiverType_01()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Is this issue specific to methods? Would it make sense to also test properties or fields?

**Comment by jcouv** [comment]
Should this PR target `dev16.0-preview2` instead?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: There are a number of existing tests that rely on the update in `VisitMemberAccess`, but I've added explicit tests here.

**Comment by cston** [comment]
See https://github.com/dotnet/roslyn/issues/30946.

**Comment by cston** [comment]
@dotnet/roslyn-compiler for a second review.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -27888,6 +27888,66 @@ void Test1(Test? x1)
...
+        b2.E += f3;
+        b2.E += f4;
+    }
+    static void M3()
+    {
+        var b3 = new B();
+        b3.E += (A<object> a) => { }; // 5
+        b3.E += (A<object>? a) => { }; // 6
+        b3.E += (A<object?> a) => { };
+        b3.E += (A<object?>? a) => { }; // 7
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: This seems unfortunate, but given that this is already how variance works for lambdas/method groups, I don't see a way to change this without making overload resolution rules extremely convoluted and inconsistent.

**Comment by cston** [comment]
@dotnet-bot test this please

**Comment by cston** [comment]
@dotnet-bot test this please

**Comment by cston** [comment]
@jaredpar for approval

**Comment by cston** [comment]
@dotnet-bot test roslyn-CI

**Comment by jaredpar** [comment]
approved

**Pull request merged**